### ISSUE TITLE
DVS-Gesture Jupyter  Visualize the spike data is black

### ISSUE THREAD

**Issue by fourmia** [bamsumit/slayerPytorch#80]
when i use dvs(.npy) dataset and IBMGestureDataset class #54 , Visualize the input spikes (first five samples) draw image is black, i use snn.io.showTD(snn.io.spikeArrayToEvent(input.reshape((2, 128, 128, -1)).cpu().data.numpy())), what should i do？

**Comment by bamsumit** [comment]
are you sure you are using `snn.io.readNpSpikes` instead of `snn.io.read2DSpikes` to read the spikes?

**Comment by fourmia** [comment]
> 

yes，I use it,
class IBMGestureDataset(Dataset):
    def __init__(self, datasetPath, sampleFile, samplingTime, sampleLength):
        self.path = datasetPath
        self.samples = pd.read_csv(sampleFile)
        self.samplingTime = samplingTime
        self.nTimeBins    = int(sampleLength / samplingTime)

    def __getitem__(self, index):
        # Read inoput and label
        # inputIndex  = self.samples[index, 0]
        # classLabel  = self.samples[index, 1]
        inputIndex = self.samples['sample'][index]
        classLabel = self.samples['labels'][index]
        # Read input spike
        inputSpikes = snn.io.readNpSpikes(
                        self.path + str(inputIndex) + '.npy'
                        ).toSpikeTensor(torch.zeros((2,128,128,self.nTimeBins)),
                        samplingTime=self.samplingTime)
        # Create one-hot encoded desired matrix
        desiredClass = torch.zeros((11, 1, 1, 1))
        desiredClass[classLabel,...] = 1

        return inputSpikes, desiredClass, classLabel

    def __len__(self):
        return self.samples.shape[0]

**Comment by bamsumit** [comment]
I am assuming you are running python script. What is the source of the dataset?

**Comment by fourmia** [comment]
> I am assuming you are running python script. What is the source of the dataset?
I use #54  offerd numpy format dataset, and use it train.txt、test.txt, I  can train model, and i use this code can show image:
for i in range(5):
    input, target, label = trainingSet[i]
    plt.imshow(((input.reshape((2, 128, 128, -1)).cpu().data.numpy())[0,...,0]).swapaxes(0,1))

**Comment by fourmia** [comment]
> I am assuming you are running python script. What is the source of the dataset?

in Train the network part, i get
Epoch :          0,  224248.6190 ms elapsed
loss = 16.909                             accuracy = 0.20427                  
loss = 13.479                             accuracy = 0.15909         
Then i kill it

**Comment by fourmia** [comment]
> I am assuming you are running python script. What is the source of the dataset?

I read the source code, In spikeArrayToEvent, i understand xEvent is image width, yEvent is image hight, next call _showTD2D, imshow image and seted it data, but i can not fix it

**Comment by fourmia** [comment]
I find snn.io.showTD return is None, Is it correct? Looking forward to your reply, Thanks a lot!

**Comment by fourmia** [comment]
> I am assuming you are running python script. What is the source of the dataset?

an example In  npy dataset, I find p max is 1, in this case, I call snn.io.showTD , Can it cause bugs or crashes?

**Comment by bamsumit** [comment]
Have you tried plotting the data directly instead of investigating inside `showTD`? What makes you think `showTD` is the problem?

```
event = snn.io.readNpSpikes('<sample.npy>')
ax = plt.axes(projection='3d')
ax.scatter3D(event.x, event.y, event.t, c=event.p);
```
and compare it to
```
data = np.load('<sample.npy>')
ax = plt.axes(projection='3d')
ax.scatter3D(data[:, 0], data[:, 1], data[:, -1], c=data[:, 2]);
```

**Comment by fourmia** [comment]
> Have you tried plotting the data directly instead of investigating inside `showTD`? What makes you think `showTD` is the problem?
> 
> ```
> event = snn.io.readNpSpikes('<sample.npy>')
> ax = plt.axes(projection='3d')
> ax.scatter3D(event.x, event.y, event.t, c=event.p);
> ```
> 
> and compare it to
> 
> ```
> data = np.load('<sample.npy>')
> ax = plt.axes(projection='3d')
> ax.scatter3D(data[:, 0], data[:, 1], data[:, -1], c=data[:, 2]);
> ```

I am not sure, I try to print a sample(user15_fluorescent_led/6.npy), I found it is (xxx,4) shape,x and y range is 0-127, p range is 0-1, t is 0.028-500+, I find use event and data plot is same

**Comment by fourmia** [comment]
This is my result;
![Snipaste_2021-12-11_09-17-40](https://user-images.githubusercontent.com/26623771/145658705-7c0e35eb-1b81-4808-a87e-4f4774008cb4.jpg)
![Snipaste1](https://user-images.githubusercontent.com/26623771/145658766-e99c18a7-e416-427d-be3d-49a44bb6ce49.jpg)
![2](https://user-images.githubusercontent.com/26623771/145658774-c3fda764-1851-4a24-a7b6-1b8843e4ff48.jpg)

**Comment by bamsumit** [comment]
Try this. `showTD` only shows the last frame in jupyter notebook.

```python
import IPython.display as display
event = snn.io.spikeArrayToEvent(input.reshape((2, 128, 128, -1)).cpu().data.numpy())
display.HTML(snn.io.animTD(event))
```

If that does not work, try it with `event = snn.io.readNpSpikes('<sample.npy>')`. If this works and previous doesn't there is something wrong with your dataloader code.

**Comment by fourmia** [comment]
> Try this. `showTD` only shows the last frame in jupyter notebook.
> 
> ```python
> import IPython.display as display
> event = snn.io.spikeArrayToEvent(input.reshape((2, 128, 128, -1)).cpu().data.numpy())
> display.HTML(snn.io.animTD(event, repeat=True))
> ```
> 
> If that does not work, try it with `event = snn.io.readNpSpikes('<sample.npy>')`. If this works and previous doesn't there is something wrong with your dataset code.

It show 'FuncAnimation' object is not subscriptable

**Comment by fourmia** [comment]
> Try this. `showTD` only shows the last frame in jupyter notebook.
> 
> ```python
> import IPython.display as display
> event = snn.io.spikeArrayToEvent(input.reshape((2, 128, 128, -1)).cpu().data.numpy())
> display.HTML(snn.io.animTD(event, repeat=True))
> ```
> 
> If that does not work, try it with `event = snn.io.readNpSpikes('<sample.npy>')`. If this works and previous doesn't there is something wrong with your dataset code.

I save gif, it is show that
![test](https://user-images.githubusercontent.com/26623771/145667857-1f3a4c1a-76b5-4a33-ac51-af32aa9117c4.gif)

**Comment by bamsumit** [comment]
It looks like your input contains multiple batches of data and you are interpreting it incorrectly.
